_posts/2025-01-16-talking-tock-58.md

```diff
@@ -0,0 +1,42 @@
+---
+title: Talking Tock 58
+subtitle: New pointer type in the kernel
+authors: bradjc
+---
+
+Correctly expressing data types in Rust between userspace and the kernel with
+correctly captured semantics is challenging as different hardware platforms
+provide different data representations. The common example of this is 32-bit
+platforms versus 64-bit platforms which store different sized types in registers
+and therefore provide a different amount of data between userspace and the
+kernel. Emerging platforms, for example hardware with CHERI support, complicate
+this further. With [pull request #4174](https://github.com/tock/tock/pull/4174)
+merged, Tock has taken a step to improve this by adding `CapabilityPtr`, a type
+within the kernel to clearly express when data within the kernel is a pointer.
+
+The `CapabilityPtr` Type
+------------------------
+
+Fundamentally this type stores a pointer to memory. The name "capability"
+signifies this type also captures the validity of that pointer and whether the
+holder of the pointer can actually use the pointer to access memory.
+
+Traditionally, a pointer in `unsafe` Rust always has the capability to be
+dereferenced. That is, the hardware will access the referenced memory.
+`CapabilityPtr` expands this abstraction, and enables software to track whether
+hardware will permit the memory access, if the hardware has such restrictions.
+
+Support for CHERI
+-----------------
+
+The primary driver for adding the `CapabilityPtr` type is the
+[CHERI](https://www.cl.cam.ac.uk/research/security/ctsrd/cheri/) extension for
+hardware architectures. With CHERI, hardware can track whether a pointer can be
```

Comment:
When someone talks about running "Rust on CHERI", they're *usually* referring to a pure-capability CHERI ABI. However, @LawrenceEsswood is porting the Tock kernel to a hybrid CHERI ABI, which has some different properties. One of those properties is that in hybrid CHERI, a language's default pointer types are not CHERI capabilities -- which is the reason why we need `CapabilityPtr`. In purecap CHERI, we don't need `CapabilityPtr`, as `*mut T` can do the same job.

We should probably clarify that `CapabilityPtr` is to support hybrid CHERI, otherwise this may be confusing for readers familiar with the upstream efforts to port Rust to CHERI.

(Lawrence, please correct me if I'm wrong about this).

Reply:
I can try to clarify this, but maybe this indicates this post is helpful as this difference is pretty opaque to me.

Reply:
Jonathan is correct. At the end of the day both `*const T` and `CapabilityPtr` are types in an abstract machine, and what hardware representation they have is always with respect to a given ABI.

"Legacy" CHERI ABI says that these should be lowered to an integer address
Hybrid CHERI ABIs say that `*const T` should be lowered to an address but `CapabilityPtr` to a CHERI capability.
Purecap CHERI ABIs say that both should be lowered to a CHERI capability.